### PR TITLE
CircleCI: Adds pylint test for all python files in scripts/ directory.

### PR DESCRIPTION
**Pull request by christianparpart**
I think it is worth it, but it initially generates tons of errors (maybe we want to tweak the pylintrc some more).

### CODE REVIEW

**Comment by axic** [comment]
@christianparpart it fails on CI:
> /usr/bin/python3: No module named pip

File: scripts/splitSources.py [hunk 1]
```diff
@@ -27,8 +27,8 @@ def extractSourceName(line):
 # writes the following source into a file named sourceName
 def writeSourceToFile(lines):
     filePath, srcName = extractSourceName(lines[0])
-    # print "sourceName is", srcName
-    # print "filePath is", filePath
+    # print("sourceName is ", srcName)
+    # print("filePath is", filePath)
```
Comment (erak) [inline scripts/splitSources.py:31]: Wouldn't this then also need an additional whitespace?

File: scripts/pylintrc [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    p,
+    x
+
+[FORMAT]
+
+# Expected format of line ending, e.g. empty (any line ending), LF or CRLF.
+expected-line-ending-format=LF
+
+# Maximum number of characters on a single line.
+max-line-length=130
```
Comment (erak) [inline scripts/pylintrc]: Perhaps add a `# TODO reduce and fix scripts`?

File: scripts/pylintrc [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+# Disable the message, report, category or checker with the given id(s). You
+# can either give multiple identifiers separated by comma (,) or put this
+# option multiple times (only on the command line, not in the configuration
+# file where it should appear only once).You can also use "--disable=all" to
+# disable everything first and then re-enable specific checks. For example, if
+# you want to run only the similarities checker, you can use "--disable=all
+# --enable=similarities". If you want to run only the classes checker, but have
+# no Warning level messages displayed, use"--disable=all --enable=classes
+# --disable=W"
+
```
Comment (erak) [inline scripts/pylintrc:13]: Perhaps add a comment about the checks you would want to enable soon?

File: scripts/splitSources.py (hunk 1 above)
Comment (erak) [reply scripts/splitSources.py:31]: Ah, wait. Could these comments be removed at all?

Comment (christianparpart) [reply scripts/splitSources.py:31]: actually, no.

File: scripts/pylint_all.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
+#! /usr/bin/env python3
+
+"""
+Performs pylint on all python files in the project repo's test directory recursively.
```
Comment (Marenz) [inline scripts/pylint_all.py:4]: This comment is not correct as it also runs for `script/`. Also there is one python script in `docs/` that doesn't seem to be tested?

**Comment by Marenz** [comment]
Maybe I am missing something, but that python script feels like overkill to me... wouldn't a simple bash script suffice?

```
find test scripts doc -iname '*.py' -exec pylint --rcfile="pylintrc" \{\} \;
```

**Review by Marenz — CHANGES_REQUESTED** [review]
Not sure about that full python script instead of a bash oneliner. at the very least the comment is not correct :>

File: scripts/pylint_all.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+This script is meant to be run from the CI but can also be easily in local dev environment,
+where you can optionally pass `-d` as command line argument to let this script abort on first error.
+"""
+
+from os import path, walk, system
+from sys import argv, exit as brexit
```
Comment (erak) [inline scripts/pylint_all.py:11]: To be honest, I'd vote for remain, I mean, for rename :wink:

File: scripts/pylint_all.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            if dev_mode:
+                return 1, checked_count
+            else:
+                failed.append(filename)
+
+    return len(failed), len(filenames)
+
+def main():
+    """" Collects all python script root dirs and runs pylint on them. """
+    dev_mode = len(argv) == 2 and argv[1] == "-d"
```
Comment (erak) [inline scripts/pylint_all.py:46]: Please document.

File: scripts/pylint_all.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    checked_count = 0
+    failed = []
+    for filename in filenames:
+        checked_count += 1
+        cmdline = "pylint --rcfile=\"{}\" \"{}\"".format(PYLINT_RCFILE, filename)
+        print("{}[{}/{}] Running pylint on file: {}{}".format(SGR_INFO, checked_count, len(filenames),
+                                                              filename, SGR_CLEAR))
+        exit_code = system(cmdline)
+        if exit_code != 0:
+            if dev_mode:
```
Comment (erak) [inline scripts/pylint_all.py:37]: Is this meant to be set by a human?

File: scripts/pylint_all.py (hunk 5 above)
Comment (christianparpart) [reply scripts/pylint_all.py]: Is aexit besser?

File: scripts/pylint_all.py (hunk 7 above)
Comment (christianparpart) [reply scripts/pylint_all.py:37]: yes. it actually greatly helped me during running the script and fixing reports.

**Comment by erak** [comment]
@Marenz I think it's right on the edge of being overkill. But it comes with a nice reporting, which you wouldn't get with a simple bash one-liner. I think we would profit from this in the future, when migrating scripts from bash to python, which I would prefer. Don't know if that's wanted at all, but could be a reason to keep this PR as it is.

**Comment by erak** [comment]
@christianparpart I'm reading https://circleci.com/gh/ethereum/solidity/251823?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link. Shouldn't the error being printed red? At least this would make it easy to find. And also would justify the effort put into the reporting.

**Review by erak — CHANGES_REQUESTED** [review]
The pylint job is failing. Apart from that, I'd be fine with merging it.

File: scripts/pylint_all.py (hunk 4 above)
Comment (Marenz) [reply scripts/pylint_all.py:4]: it still says "in the repo's test directory"?

File: scripts/pylint_all.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+                failed.append(filename)
+
+    return len(failed), len(filenames)
+
+def main():
+    """" Collects all python script root dirs and runs pylint on them. """
+    dev_mode = len(argv) == 2 and argv[1] == "-d"
+    failed_count, total_count = pylint_all_filenames(dev_mode, [
+        path.abspath(path.dirname(__file__) + "/../scripts"),
+        path.abspath(path.dirname(__file__) + "/../test")])
```
Comment (Marenz) [inline scripts/pylint_all.py:49]: do we want to run this on the script in `docs` as well?

**Review by chriseth — APPROVED** [review]

**Comment by Marenz** [comment]
@chriseth  I am not sure what happened here, but it looks to me like you [reve](https://github.com/ethereum/solidity/compare/7802fba890b7ecf87bcf10b7b7c6b337297d16df..33f990a96a9ea30bb1e77c267bded7f8b4618d68#diff-69d1c3390ca19edfb5424cb13240ddc1L4)[rted](https://github.com/ethereum/solidity/compare/7802fba890b7ecf87bcf10b7b7c6b337297d16df..33f990a96a9ea30bb1e77c267bded7f8b4618d68#diff-69d1c3390ca19edfb5424cb13240ddc1L48) the fixes in your force push that I commented on?

**Comment by chriseth** [comment]
Thanks a lot for noticing! I thought this was a fresh-checkout but somehow it wasn't should be fixed in https://github.com/ethereum/solidity/pull/8239

**Pull request merged**